Sources/TuistLoader/Models+ManifestMappers/Headers+ManifestMapper.swift

```diff
@@ -12,58 +12,142 @@ extension TuistGraph.Headers {
         let privateHeaders: [AbsolutePath]
         let projectHeaders: [AbsolutePath]
 
-        func resolveHeaders(_ list: FileList?) throws -> [AbsolutePath] {
+        func unfold(_ list: FileList?, headersFromUmbrella: Set<String>? = nil) throws -> [AbsolutePath] {
             guard let list = list else { return [] }
-            return try list.globs.flatMap {
-                unfoldGlob(
-                    try generatorPaths.resolve(path: $0.glob),
-                    excluding: (try resolveExcluding($0.excluding)).union(autoExlcudedPaths)
-                )
+            var result = try list.unfold(
+                generatorPaths: generatorPaths,
+                basenameFilter: headersFromUmbrella,
+                additionalExcluding: autoExlcudedPaths
+            )
+            // be sure, that umbrella already here (if we use it)
+            if headersFromUmbrella != nil,
+               let resolvedUmbrellaPath = resolvedUmbrellaPath,
+               !result.contains(resolvedUmbrellaPath)
+            {
+                result.append(resolvedUmbrellaPath)
             }
+            return result
         }
 
         switch manifest.exclusionRule {
         case .projectExcludesPrivateAndPublic:
-            publicHeaders = try resolveHeaders(manifest.public)
+            publicHeaders = try unfold(manifest.public, headersFromUmbrella: headersFromUmbrella)
             autoExlcudedPaths.formUnion(publicHeaders)
-            privateHeaders = try resolveHeaders(manifest.private)
+            privateHeaders = try unfold(manifest.private)
             autoExlcudedPaths.formUnion(privateHeaders)
-            projectHeaders = try resolveHeaders(manifest.project)
+            projectHeaders = try unfold(manifest.project)
 
         case .publicExcludesPrivateAndProject:
-            projectHeaders = try resolveHeaders(manifest.project)
+            projectHeaders = try unfold(manifest.project)
             autoExlcudedPaths.formUnion(projectHeaders)
-            privateHeaders = try resolveHeaders(manifest.private)
+            privateHeaders = try unfold(manifest.private)
             autoExlcudedPaths.formUnion(privateHeaders)
-            publicHeaders = try resolveHeaders(manifest.public)
+            publicHeaders = try unfold(manifest.public, headersFromUmbrella: headersFromUmbrella)
         }
 
         return Headers(public: publicHeaders, private: privateHeaders, project: projectHeaders)
     }
 }
+
+extension FileList {
+    fileprivate func unfold(
+        generatorPaths: GeneratorPaths,
+        basenameFilter: Set<String>? = nil,
+        additionalExcluding: Set<AbsolutePath>
+    ) throws -> [AbsolutePath] {
+        return try globs.flatMap {
+            try $0.unfold(
+                generatorPaths: generatorPaths,
+                basenameFilter: basenameFilter,
+                additionalExcluding: additionalExcluding
+            )
+        }
+    }
+}
+
+extension FileListGlob {
```

Comment:
I'd move this one into `FileListGlob.swift` and add a unit test to it

Reply:
done, except unit-test (bit later), but anyway I made it abstract and this is just extracted func which we already have in header loader